plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Completions.hs

```diff
@@ -0,0 +1,635 @@
+  -- ^ range where completion is to be inserted
+  }
+  deriving (Eq, Show, Read)
+
+{- | The context a cursor can be in within a cabal file,
+  we can be in stanzas or the top level,
+  and additionally we can be in a context where we have already
+  written a keyword but no value for it yet
+-}
+type Context = (NestingContext, KeyWordContext)
+
+-- | Context inside a cabal file, used to decide which keywords to suggest
+data NestingContext
+  = -- | Top level context in a cabal file such as 'author'
+    TopLevel
+  | -- | Nested context in a cabal file, such as 'library',
+    -- which has nested keywords, specific to the stanza
+    Stanza StanzaName
+  deriving (Eq, Show, Read)
+
+{- | Keyword context in cabal file
+  used to decide whether to suggest values or keywords
+-}
+data KeyWordContext
+  = -- | Key word context, where a keyword
+    -- occurs right before the current position,
+    -- with no value associated to it
+    KeyWord KeyWordName
+  | -- | Keyword context where no keyword occurs
+    -- right before the current position
+    None
+  deriving (Eq, Show, Read)
+
+type KeyWordName = T.Text
+type StanzaName = T.Text
+
+{- | Information about the current completion status
+
+  Example: @"dir1/fi@ having been written to the file
+  would correspond to:
+
+  @
+    completionPrefix = "dir1/fi"
+    completionSuffix  = Just "\\""
+    ...
+  @
+-}
+data CabalCompletionContext = CabalCompletionContext
+  { completionPrefix       :: T.Text
+  -- ^ text prefix to complete
+  , completionSuffix       :: Maybe T.Text
+  -- ^ possible wrapping text, to write after
+  --   the filepath has been completed
+  , completionRange        :: Range
+  -- ^ range where completion is to be inserted
+  , completionCabalFileDir :: FilePath
+  -- ^ filepath of the handled file
+  }
+  deriving (Eq, Show, Read)
+
+-- ----------------------------------------------------------------
+-- Public API for Completions
+-- ----------------------------------------------------------------
+
+{- | Takes information about the completion status within the file
+  and finds the correct completer to be applied
+-}
+contextToCompleter :: Context -> Completer
+-- if we are in the top level of the cabal file and not in a keyword context,
+-- we can write any top level keywords or a stanza declaration
+contextToCompleter (TopLevel, None) =
+  constantCompleter $
+    Map.keys (cabalVersionKeyword <> cabalKeywords) ++ Map.keys stanzaKeywordMap
+-- if we are in a keyword context in the top level,
+-- we look up that keyword in the top level context and can complete its possible values
+contextToCompleter (TopLevel, KeyWord kw) =
+  case Map.lookup kw (cabalVersionKeyword <> cabalKeywords) of
+    Nothing -> noopCompleter
+    Just l  -> l
+-- if we are in a stanza and not in a keyword context,
+-- we can write any of the stanza's keywords or a stanza declaration
+contextToCompleter (Stanza s, None) =
+  case Map.lookup s stanzaKeywordMap of
+    Nothing -> noopCompleter
```

Comment:
Since we use the same keyword map previously to be able to recognise the stanza context for `s` in the first place, it should not be possible to not find `s` with the lookup operation. 

But since something must have gone wrong for this case to occur, I think logging is sensible.